### PR TITLE
Ticket3905 visual indicator nicos running

### PR DESCRIPTION
**Pull request by RaiBishal**
### Description of work

Added visual indicator when the script is running in script server.

### Ticket

https://github.com/ISISComputingGroup/IBEX/issues/3905

### Acceptance criteria

Script Server button to show appropriate icon and flash when the script is running

### Unit tests

*Give an overview of unit tests you have added or modified, if applicable. The aim is provide information to help the reviewer*

### System tests

*Mention any automated tests or manual tests that you have added or modified, if applicable. The aim is provide information to help the reviewer*

### Documentation
*Highlight and provide a link to any additions or changes to the documentation, if applicable. The aim is provide information to help the reviewer*

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
+package uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher.controls;
+
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+
+import org.eclipse.swt.SWT;
```
Comment (ChrisColeExpControl) [inline base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java]: Are these really needed? They're not being used and it runs fine without them.

File: base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+		} else if (status == ScriptStatus.INBREAK) {
+			flash.stop();
+			setText(Label + PauseButton);
+		} else {
+			flash.stop();
+			setText(Label);
+		}
+	}
+	
+}
```
Comment (ChrisColeExpControl) [inline base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java]: Should be a new line at the end of the file.

File: base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/AlarmButtonViewModel.java [hunk 3]
```diff
@@ -33,20 +33,18 @@
...
+        alarmCounter.addPropertyChangeListener("alarmCount", arg0 -> {
                 alarmCount = alarmCounter.getCount();
                 if (!active) {
                     alarmSeen = false;
                 }
                 update();
                 updateFlashing();
             }
-        });
+        );
```
Comment (ChrisColeExpControl) [inline base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/AlarmButtonViewModel.java]: Confusing indentation, did you mean to delete the other bracket?

File: base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	/**
+	 * 
+	 * @param model 
+	 * 			the model that ScriptServerButton listens to 
+	 * @param buttonLabel
+	 * 			the label to display on the button
+	 */
+	public ScriptServerButtonViewModel(NicosModel model, String buttonLabel) {
+		super(buttonLabel);
+		// TODO Auto-generated constructor stub
```
Comment (Tom-Willemsen) [inline base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java]: Remove TODO

File: base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		flash = new FlashingButton(this, SCRIPT_BUTTON_COLOR);
+		flash.setDefaultColour(DEFOCUSSED);
+		Label = buttonLabel;
+		
+		model.addPropertyChangeListener("scriptStatus", arg0->{
```
Comment (Tom-Willemsen) [inline base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java]: Minor readability suggestion

```suggestion
		model.addPropertyChangeListener("scriptStatus", ignored->{
```

File: base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/AlarmButtonViewModel.java [hunk 6]
```diff
@@ -33,20 +33,18 @@
...
         super(buttonLabel);
 
-        flash = new FlashingButton(this);
+        flash = new FlashingButton(this, ALARM_COLOR);
         flash.setDefaultColour(DEFOCUSSED);
 
-        alarmCounter.addPropertyChangeListener("alarmCount", new PropertyChangeListener() {
-            @Override
-            public void propertyChange(PropertyChangeEvent arg0) {
+        alarmCounter.addPropertyChangeListener("alarmCount", arg0 -> {
```
Comment (Tom-Willemsen) [inline base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/AlarmButtonViewModel.java]: Minor readability suggestion
```suggestion
        alarmCounter.addPropertyChangeListener("alarmCount", ignored -> {
```

File: base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java (hunk 1 above)
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java]: Removed.

File: base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java (hunk 2 above)
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java]: Done.

File: base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/AlarmButtonViewModel.java (hunk 3 above)
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/AlarmButtonViewModel.java]: Fixed.

File: base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java (hunk 4 above)
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java]: Done.

File: base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/AlarmButtonViewModel.java (hunk 6 above)
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/AlarmButtonViewModel.java]: Done.

File: base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java (hunk 5 above)
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.e4.ui.perspectiveswitcher/src/uk/ac/stfc/isis/ibex/e4/ui/perspectiveswitcher/controls/ScriptServerButtonViewModel.java]: Done.

**Comment by RaiBishal** [comment]
@ChrisColeExpControl Changes have been implemented. 

- Green steady non-flashing button when the script is running and the button is defocused(with play icon)
- Green flashing button when the script is paused when the button is defocused(with paused icon)
- Blue button when the script is running/paused and the button is focused(appropriate icon displayed)

**Comment by ChrisColeExpControl** [comment]
When the user hovers their mouse over the button while a script is running, the color does not display correctly (should be either green or light blue)

**Pull request merged**